### PR TITLE
Establish p2p channel via broker

### PR DESCRIPTION
**Pull request by anjmao**
1. Key and config exchange for p2p via broker.
2. Migrated p2p messages to protobuf.
3. P2P channel rewritten with raw kcp library UDP session.

**Consumer will not use p2p channel yet**, so old flow should work as is, but I want to review and merge it as we are getting merge conflicts.

**TODO for next PRs:**
1. Pass service type to channel creation so it can work for multiple services (Provider side).
2. Enable p2p channel on consumer side.
3. Pass second punched UDP conn to services (Consumer side).
4. Add internal message send retry.

### CODE REVIEW

File: localnet/provider.sh [hunk 1]
```diff
@@ -28,4 +28,4 @@ exec /node/build/myst/myst \
   --identity=0xd1a23227bd5ad77f36ba62badcb78a410a1db6c5 \
   --identity.passphrase=localprovider \
   --openvpn.port=3000 \
-  openvpn,wireguard
+  wireguard
```
Comment (anjmao) [inline localnet/provider.sh]: Revert this change.

File: core/service/subscription.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		consumerID := identity.FromAddress(si.GetConsumerID())
+		sessionID := si.GetSessionID()
+
+		err := mng.Acknowledge(consumerID, sessionID)
+		if err != nil {
+			log.Error().Err(err).Msgf("Cannot acknowledge session %s", sessionID)
+			return err
+		}
+
+		return nil
```
Comment (anjmao) [inline core/service/subscription.go]: Probably should be `return c.Ok()`

File: core/service/subscription.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	ch.Handle(p2p.TopicSessionDestroy, func(c p2p.Context) error {
+		var si pb.SessionInfo
+		proto.Unmarshal(c.Request().Data, &si)
+
+		consumerID := identity.FromAddress(si.GetConsumerID())
+		sessionID := si.GetSessionID()
+
+		err := mng.Destroy(consumerID, sessionID)
+		if err != nil {
+			log.Error().Err(err).Msgf("Cannot destroy session %s", sessionID)
```
Comment (anjmao) [inline core/service/subscription.go]: No need to log in handlers. Could be simple fmt.Errorf.
```go
if err != nil {
     return fmt.Errorf("cannot destroy session %s: %v", sessionID, err)
}
```
Internally handler can log in one place with added context like which failed topic.

File: core/service/subscription.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+		data, err := json.Marshal(config.SessionServiceConfig)
+		if err != nil {
+			log.Error().Err(err).Msgf("Cannot pack session %s service config", string(session.ID))
+			return err
+		}
+
+		pc := p2p.ProtoMessage(&pb.SessionResponse{
+			ID:          string(session.ID),
+			DNSs:        "",
```
Comment (soffokl) [inline core/service/subscription.go]: This DNS is not filled here, we should add it.

File: core/service/manager.go [hunk 5]
```diff
@@ -137,6 +145,17 @@ func (manager *Manager) Start(providerID identity.Identity, serviceType string,
...
 
+	err = manager.p2pManager.SubscribeChannel(providerID, func(ch *p2p.Channel) {
+		mng := manager.sessionManager(proposal, string(id), ch)
+		subscribeSessionCreate(mng, ch, service, id)
+		subscribeSessionAcknowledge(mng, ch)
+		subscribeSessionDestroy(mng, ch)
+	})
+
+	if err != nil {
+		return id, err
```
Comment (zolia) [inline core/service/manager.go]: wrap an error.

File: p2p/manager.go [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+			log.Err(err).Msg("Could not handle exchange ack")
+			return
+		}
+
+		// Send ack in separate goroutine and start pinging.
+		// It is important that provider starts sending pings first otherwise
+		// providers router can think that consumer is sending DDoS packets.
+		go func(reply string) {
+			if err := brokerConn.Publish(reply, []byte("OK")); err != nil {
+				log.Err(err).Msg("Could publish exchange ack")
```
Comment (zolia) [inline p2p/manager.go]:
```suggestion
				log.Err(err).Msg("Could not publish exchange ack")
```

File: p2p/manager.go [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+func (m *Manager) providerStartConfigExchange(brokerConn nats.Connection, signerID identity.Identity, msg *nats_lib.Msg) error {
+	pubKey, privateKey, err := GenerateKey()
+	if err != nil {
+		return fmt.Errorf("could not generate provider p2p keys: %w", err)
+	}
+
+	// Get initial peer exchange with it's public key.
+	signedMsg, err := m.unpackSignedMsg(msg.Data)
+	if err != nil {
+		return err
```
Comment (zolia) [inline p2p/manager.go]: wrap errors.

File: session/pingpong/exchange_messaging.go [hunk 8]
```diff
@@ -33,18 +35,38 @@ const messageEndpointExchange = communication.MessageEndpoint(endpointExchange)
...
 // NewExchangeSender returns a new instance of exchange message sender.
-func NewExchangeSender(sender communication.Sender) *ExchangeSender {
+func NewExchangeSender(sender communication.Sender, ch *p2p.Channel) *ExchangeSender {
 	return &ExchangeSender{
+		ch:     ch,
 		sender: sender,
 	}
 }
 
 // Send send the given exchange message.
```
Comment (zolia) [inline session/pingpong/exchange_messaging.go]:
```suggestion
// Send sends the given exchange message.
```

File: session/pingpong/factory.go [hunk 9]
```diff
@@ -137,6 +138,43 @@ func InvoiceFactoryCreator(
...
+		if err := dialog.Receive(listener.GetConsumer()); err != nil {
+			return nil, err
+		}
+
+		return exchangeChan, nil
+	}
+
+	channel.Handle(p2p.TopicPaymentMessage, func(c p2p.Context) error {
+		var msg pb.ExchangeMessage
+		proto.Unmarshal(c.Request().Data, &msg)
```
Comment (zolia) [inline session/pingpong/factory.go]: swallowed err?

File: session/pingpong/factory.go [hunk 10]
```diff
@@ -185,3 +221,32 @@ func ExchangeFactoryFunc(
...
+		err := dialog.Receive(listener.GetConsumer())
+		if err != nil {
+			return nil, err
+		}
+		return invoices, nil
+	}
+
+	channel.Handle(p2p.TopicPaymentInvoice, func(c p2p.Context) error {
+		var msg pb.Invoice
+		proto.Unmarshal(c.Request().Data, &msg)
```
Comment (zolia) [inline session/pingpong/factory.go]: swallowed error.

File: p2p/channel.go [hunk 11]
```diff
@@ -18,106 +18,184 @@
...
+		return nil, fmt.Errorf("could not create UDP session: %w", err)
 	}
-	c.initPeer()
 
-	ln, err := c.listen()
-	if err != nil {
-		return nil, err
+	c := &Channel{
+		conn:                  textproto.NewConn(udpSession),
+		handlers:              make(map[string]HandlerFunc),
```
Comment (zolia) [inline p2p/channel.go]: would suggest naming: messageHandlers, topicHandlers or requestHandlers

File: p2p/channel.go [hunk 12]
```diff
@@ -18,106 +18,184 @@
...
+}
+
+// handleReply forwards reply message to associated stream result channel.
+func (c *Channel) handleReply(msg *transportMsg) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	if s, ok := c.streams[msg.id]; ok {
+		s.resCh <- msg
+	} else {
+		log.Warn().Msgf("Stream %d not found", msg.id)
```
Comment (zolia) [inline p2p/channel.go]: I would log message itself for missing stream. Would help while debugging.

File: p2p/channel.go [hunk 13]
```diff
@@ -18,106 +18,184 @@
...
+		resMsg.statusCode = statusCodeOK
+		if ctx.res != nil {
+			resMsg.data = ctx.res.Data
+		}
+	}
+	c.sendQueue <- &resMsg
+}
+
+// ServiceConn returns UDP connection which can be used for services.
+func (c *Channel) ServiceConn() *net.UDPConn {
```
Comment (zolia) [inline p2p/channel.go]: a bit unclear why we have this method here? I thought traversal package should handle this.

File: p2p/channel.go [hunk 14]
```diff
@@ -153,107 +223,50 @@ func (c *Channel) Handle(topic string, handler HandlerFunc) {
...
-
-	ctx := defaultContext{req: &Message{Data: reqData}}
-	err = handler(&ctx)
-
-	if err != nil {
-		log.Err(err).Msg("Internal handler error")
-		w.WriteHeader(http.StatusInternalServerError)
-		return
-	}
+// sendRequest sends data bytes via HTTP POST request and optionally reads response body.
```
Comment (zolia) [inline p2p/channel.go]: old comment

File: p2p/channel.go [hunk 15]
```diff
@@ -267,3 +280,18 @@ func newBlockCrypt(privateKey PrivateKey, peerPublicKey PublicKey) (kcp.BlockCry
...
+	rawConn.Close()
+
+	raddr := rawConn.RemoteAddr().(*net.UDPAddr)
+	network := "udp4"
+	if raddr.IP.To4() == nil {
+		network = "udp"
+	}
+	conn, err := net.ListenUDP(network, rawConn.LocalAddr().(*net.UDPAddr))
+	if err != nil {
+		return nil, err
```
Comment (zolia) [inline p2p/channel.go]: wrap an error

File: p2p/channel.go (hunk 13 above)
Comment (anjmao) [reply p2p/channel.go]: Yes, traversal package does actual pinging. This field is not used yet, not sure yet how will second conn be passed to actual service, will remove it in separate PR.

File: session/dto.go [hunk 16]
```diff
@@ -49,7 +49,7 @@ type Session struct {
 	DataTransferred DataTransferred
 	TokensEarned    uint64
 	Last            bool
-	done            chan struct{}
+	Done            chan struct{}
```
[the remark this answers is not in the document]
Comment (anjmao) [reply session/dto.go]: Exposed as read only channel.

File: core/service/subscription.go (hunk 4 above)
Comment (anjmao) [reply core/service/subscription.go]: Added to separate task https://github.com/mysteriumnetwork/node/issues/1932
